docs/zksync-era-integration.md

```diff
@@ -0,0 +1,184 @@
+
+## Operator/sequencer
+
+Currently, the operator is a centralized server (there are plans to make a decentralized consensus for operators) which can be thought of as the node entry point, its responsibilities are:
+
+-   Initializing the `era_vm` and keeping its state.
+-   Orchestrating the Bootloader and keeping its state.
+-   Keeping storage database and committing changes.
+
+Here is a simplified version of what the vm on the operator end looks like:
+
+```rust
+struct OperatorVm {
+    pub(crate) inner: EraVM, // this would be the actual `era_vm`
+    pub suspended_at: u16, // last pc when execution stopped because of a hook
+    pub Bootloader_state: BootloaderState,
+    pub(crate) storage: StorageDb,
+    pub snapshot: Option<VmSnapshot>,
+}
+```
+
+### Initializing the vm
+
+VM initialization involves:
+
+-   Loading the Bootloader bytecode and initializing its state.
+-   Setting up the `era_vm` by injecting the Bootloader code, loading default contracts, and configuring other settings.
+
+When setting up `era_vm`, the operator provides access to the chain storage database with the following API:
+
+```rust
+trait Storage {
+    fn decommit(&mut self, hash: U256) -> Option<Vec<U256>>;
+
+    fn storage_read(&mut self, key: &StorageKey) -> Option<U256>;
+
+    fn cost_of_writing_storage(&mut self, key: &StorageKey, value: U256) -> u32;
+
+    fn is_free_storage_slot(&self, key: &StorageKey) -> bool;
+}
+```
+
+This storage is saved in the VM state as a pointer. Here’s a brief explanation of each function:
+
+-   **decommit**: given a hash it returns a contract bytecode from the database.
+-   **storage_read**: given a key, it returns the potential value from the database.
+-   **cost_of_writing_storage**: when writing to the contract storage, gas is consumed, but the cost of writing depends on whether the write is initial or not. More on that [here](https://github.com/lambdaclass/zksync-era/blob/era_vm_integration_v2/docs/specs/zk_evm/fee_model.md).
+-   **is_free_storage_slot**: if the address to write belongs to the system context and the key belongs to the base L2 token address, then the storage_slot is free(doesn't incur gas charges).
+
+A few notes about this storage:
+
+1. The key has the following structure:
+
+```rust
+struct StorageKey {
+    pub address: H160,
+    pub key: U256,
+}
+```
+
+This allows us to query a key that belongs to the current executing address.
+
+2. There isn't any consensus or spec about how storage should be implemented. We came up with this API because it is what we thought was more convenient for the requirement. But, for example, the vm1 implements a query logic, where the operator will react based on the provided params.
```

Comment:
```suggestion
2. There isn't any consensus or spec about how storage should be implemented. We came up with this API because it is what we thought was more convenient for the requirement. But, for example, the vm1 implements a query logic, where the operator will react based on the [provided params](https://github.com/matter-labs/zksync-era/blob/87768755e8653e4be5f29945b56fd05a5246d5a8/core/lib/types/src/zk_evm_types.rs#L17-L30).
```